src_multispec/MassFluxUtil.cpp

```diff
@@ -1,26 +1,48 @@
 #include "multispec_functions.H"
 
-void ComputeMolconcMolmtot(const MultiFab& rho,
-			   const MultiFab& rhotot,
-			   MultiFab& molarconc,
-			   MultiFab& molmtot)
+void ComputeMolconcMolmtot(const MultiFab& rho_in,
+			   const MultiFab& rhotot_in,
+			   MultiFab& molarconc_in,
+			   MultiFab& molmtot_in)
 {
 
     BL_PROFILE_VAR("ComputeMolconcMolmtot()",ComputeMolconcMolmtot);
 
-    int ng = molarconc.nGrow();
+    int ng = molarconc_in.nGrow();
         
     // Loop over boxes
-    for (MFIter mfi(rho,TilingIfNotGPU()); mfi.isValid(); ++mfi) {
+    for (MFIter mfi(rho_in,TilingIfNotGPU()); mfi.isValid(); ++mfi) {
 
         // Create cell-centered box
         const Box& bx = mfi.growntilebox(ng);
 
-        compute_molconc_molmtot(ARLIM_3D(bx.loVect()), ARLIM_3D(bx.hiVect()),
-				BL_TO_FORTRAN_ANYD(rho[mfi]),
-				BL_TO_FORTRAN_ANYD(rhotot[mfi]),
-				BL_TO_FORTRAN_ANYD(molarconc[mfi]),
-				BL_TO_FORTRAN_ANYD(molmtot[mfi]));
+        const Array4<const Real>& rho = rho_in.array(mfi);
+        const Array4<const Real>& rhotot = rhotot_in.array(mfi);
+        const Array4<      Real>& molarconc = molarconc_in.array(mfi);
+        const Array4<      Real>& molmtot = molmtot_in.array(mfi);
+        
+
+        amrex::ParallelFor(bx, [=] AMREX_GPU_DEVICE (int i, int j, int k) noexcept
+        {
+
+            GpuArray<Real, MAX_SPECIES> W;
+
+            // calculate mass fraction and total molar mass (1/m=Sum(w_i/m_i))
+
+            for (int n=0; n<nspecies; ++n){
+                W[n] = rho(i,j,k,n) / rhotot(i,j,k);
+            }
+
+            GetMolTot(W, molmtot(i,j,k));
+
+            // calculate molar concentrations in each cell (x_i=m*w_i/m_i) 
+
+            for (int n=0; n<nspecies; ++n){
+                molarconc(i,j,k,n) = molmtot(i,j,k) * W[n] / molmass[n];
+            }
+    
+        });
+
```

Comment:
This should be put in a  separate inline function at the end of multispec_functions.H
The reason is that there are other routines that need this kernel with values of nspecies that are different from the default value from the namespace.  So nspecies should be an input, and from this particular function you want to pass in "nspecies".